rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,235 @@
+`SimpleDirectMarkerBasedEarlyConflictDetectionStrategy` for DirectWriteMarkers to perform corresponding conflict detection and
+conflict resolution. And we provide `AsyncTimelineMarkerEarlyConflictDetectionStrategy
+` for TimelineServerBasedWriteMarkers to perform corresponding conflict detection and conflict resolution
+
+#### DirectWriteMarkers related strategy
+
+##### SimpleTransactionDirectMarkerBasedEarlyConflictDetectionStrategy
+
+![](figure2.png)
+
+As for current strategy showed by figure 2, we need to pay attention to the following two details:
+
+
+Firstly, we will use the existing transaction mechanism to lock at the <strong>PartitionPath + "/" + fileID</strong>
+level before checking conflicts and before creating marker files. 
+Take ZK locker as an example => `LockConfiguration.ZK_LOCK_KEY_PROP_KEY, partitionPath + "/" + fileId`. 
+That is to say, for the same partitionPath/fileId, only one writer can check and create the corresponding marker file 
+at the same time.
+
+
+In addition, during checking conflicts, we do not need to list all marker files in all directories, we can prune based 
+on the current partition Path and fileID to avoid unnecessary list operations. For example(also shown in figure2), we
+are going to create maker file based on partition path 2022/07/01, and fileID ff26cb9e-e034-4931-9c59-71bac578f7a0-0.
+During marker conflict checking, we do not need to list all the partitions we have under ".temp". Just list marker files under
+$BasePath/.hoodie/.temp/instantTime/2022/07/01 and check if fileID ff26cb9e-e034-4931-9c59-71bac578f7a0-0 existed or not.
+
+##### SimpleDirectMarkerBasedEarlyConflictDetectionStrategy
+
+Compared with `SimpleTransactionDirectMarkerBasedEarlyConflictDetectionStrategy`, the current strategy drops the steps of
+transaction including new transaction, begin transaction and end transaction. The advantages are that the checking
+speed is faster, and multiple writers will not affect each other. But the downside is that the conflict detection may be
+delayed.
+
+
+For example, when two writers detect PartitionPath1 + fileID1 at the same time, 
+both writers will pass conflict checking and successfully create a marker, which conflict detection is delayed.
+If so, the conflict can only be found in the pre-commit conflict checking and fail writing.
+This leads to waste of resources, but don't compromise the correctness of the data.
+
+
+As we can see, all these direct based early conflict detection strategy need extra fs calling.
+
+#### TimelineServerBasedWriteMarkers related strategy
+
+##### AsyncTimelineMarkerEarlyConflictDetectionStrategy
+
+This design expands the create marker api on timeline server.
+
+![](figure3.png)
+
+As shown on figure3. For the client side, it will call create marker api, requesting to create markers.
+
+For the timeline server side, there is a MarkerCheckerRunnable thread, which will check the conflicts between current 
```

Comment:
`For the timeline server side, there is a MarkerCheckerRunnable thread, which will check the conflicts between current writer and other active writers at fixed rate.`
->
`On the timeline server side, there is a MarkerCheckerRunnable thread, which checks the conflicts between the markers in memory created by the current writer and the markers persisted to the storage created by other active writers periodically.`
We should mention where the makers are stored for this writer vs other concurrent writers.